File: summarize_csv.rb
#!/usr/bin/env ruby
require 'csv'
require 'optparse'
require 'rbconfig'

THIS_DIR = File.expand_path(File.dirname(__FILE__))
RUBY = File.join(RbConfig::CONFIG['bindir'], RbConfig::CONFIG['ruby_install_name'])


Encoding.default_external = Encoding::UTF_8
Encoding.default_internal = Encoding::UTF_8

options = {}

optparse = OptionParser.new do |opts|
  opts.banner = "USAGE: summarize_csv.rb -f /path/to/file.csv -o /path/to/images -d /path/to/output/folder"

  opts.on("-f", "--file FILE", "CSV file") do |f|
    options[:csv] = f
  end

  opts.on("-o", "--origin_folder ORIGIN_FOLDER", "Folder where images are located") do |o|
    options[:origin] = o
  end

  opts.on("-d", "--destination DESTINATION", "Destination folder") do |d|
    options[:destination] = d
  end

  opts.on("-p", "--processes NUM_PROCESSES", "Num of CPU processes to use (default 16)") do |d|
    options[:cpu] = d
  end

  opts.on("-h", "--help", "Prints this help") do
    puts opts
    exit
  end
end

begin
  optparse.parse!
  mandatory = [:csv, :origin, :destination]
  missing = mandatory.select{ |param| options[param].nil? }
  unless missing.empty?
    raise OptionParser::MissingArgument.new(missing.join(', '))
  end
rescue OptionParser::InvalidOption, OptionParser::MissingArgument
  puts $!.to_s
  puts optparse
  exit
end

CPU = options[:cpu] || 16
csv_file = options[:csv]
origin = options[:origin]
destination = options[:destination]

if !File.file?(csv_file)
  puts "File #{csv_file} does not exist."
  exit
end  

if !File.directory?(destination)
  puts "Directory #{destination} does not exist."
  exit
end

data = CSV.parse(File.read(csv_file), headers: true)

count = data.length

group_size = (count / CPU).ceil + 1

groups = data.each_slice(group_size).to_a

threads = []

groups.each do |g|
  len = g.length
  threads << Thread.new do
    len.times do |i|
      item = g[i]
      if (!item["file_key"].nil? && !item["access_pid"].nil?)
        file_key = item["file_key"].gsub(/"/, '')
        access_pid = item["access_pid"].gsub(/"/, '')
        image = "#{origin}/#{file_key}"
        json = "#{destination}/#{access_pid}.json"
        system(RUBY, "#{THIS_DIR}/summarize_file.rb", "-i", image, "-d", json)
      end
    end
  end
  threads.map(&:join)
end

File: summarize_csv.py
#!/usr/bin/env python
import time
import pandas as pd
import subprocess
import multiprocessing
import tqdm
import argparse
import os
from pathlib import Path

parser = argparse.ArgumentParser()

parser.add_argument("-f", "--file", default="/usr/src/app/files/files-urls.csv", type=str, help="path to CSV file (default: /usr/src/app/files/files-urls.csv)")
parser.add_argument("-o", "--origin_folder", default="/usr/src/app/files", type=str, help="Folder where images are located (default: /usr/src/app/files)")
parser.add_argument("-d", "--destination", default="/usr/src/app/files/colors_output", type=str, help="Folder where to save results (default: /usr/src/app/files/colors_output)")
parser.add_argument('--overwrite', action='store_true', help='Overwrite existing json if it exists')
parser.add_argument("-p", "--processes", default=None, type=int, help="How many processes (default: None which lets Python decide)")

args = parser.parse_args()

cpu = args.processes
file = args.file
origin_folder = args.origin_folder
destination = args.destination
overwrite = args.overwrite

count=0
starttime = time.time()

if (not os.path.exists(file)):
  print("File %s does not exist." % file)
  exit()

if (not os.path.exists(origin_folder)):
  print("Folder %s does not exist." % origin_folder)
  exit()

if (not os.path.exists(destination)):
  print("Folder %s does not exist." % destination)
  exit()

url_df = pd.read_csv(file, index_col="access_pid")

count = len(url_df.filename)

skipped = []

print("Summarizing %s files with %s processes" % (count, (multiprocessing.cpu_count() if cpu is None else cpu)))

def summarize_row(access_pid, row):
  try:
    filename = row["filename"]
    image = "%s/%s/%s" % (origin_folder, filename[0:4], filename)
    json = "%s/%s.json" % (destination, access_pid)
    if (Path(json).exists() is False or overwrite):
      subprocess.run(["python", "summarize_file.py", "-s", image, json])
  except:
    skipped.append(row)
    pass

with multiprocessing.Pool(processes=cpu) as pool:
  for access_pid, row in tqdm.tqdm(url_df.iterrows(), total=count):
    pool.apply(summarize_row, args=(access_pid, row))
  pool.close()
  pool.join()

print("Processed %s files in {} seconds".format(time.time() - starttime) % count)

if (len(skipped) > 0):
  print("Skipped %s files:", len(skipped))
  print(skipped)